### DISCUSSION TITLE
What makes it?

### DISCUSSION THREAD

**Discussion by starkovv** [Ysurac/openmptcprouter#1741] [General]
### What kind of components is OpenMPTCPRouter?

1. As far as I understand the minimum requirements are:
– client node with 2 uplinks, no public IP required
– VDS with 1 uplink and public IP

2. Both should support MPTCP in kernel.

3. What version of MPTCP should it support: `v0` or `v1`?
As far as I know MPTCP package from https://multipath-tcp.org is `v0`. While the version from https://www.kernel.org is `v1`. Both are not compatible with each other.

4. What application makes the MPTCP pipe from the client to the server? Is it `Shadowsocks`?

I'm playing around with MPTCP in my lab.
**Here is the client:**
```
# uname -a
Linux nanopi-r4s 5.11.12-rockchip64 #trunk SMP PREEMPT Thu Apr 8 19:16:55 UTC 2021 aarch64 aarch64 aarch64 GNU/Linux
```
```
# dmesg | grep MPTCP
[    1.323437] MPTCP token hash table entries: 4096 (order: 4, 98304 bytes, linear)
```
```
# sysctl -a | grep mptcp
net.ipv4.tcp_available_ulp = mptcp
net.mptcp.add_addr_timeout = 120
net.mptcp.enabled = 1
```

**And here is the server:**
```
# uname -a
Linux demo-gw1 5.11.12-051112-generic #202104071432 SMP Wed Apr 7 14:50:17 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```
```
#  dmesg | grep MPTCP
[    0.498016] MPTCP token hash table entries: 1024 (order: 2, 24576 bytes, linear)
```

```
# sysctl -a | grep mptcp
net.ipv4.tcp_available_ulp = espintcp mptcp
net.mptcp.add_addr_timeout = 120
net.mptcp.enabled = 1
```

**Comment by Ysurac** [comment]
For now I don't "play" with upstreamed MPTCP and there is lot of commit in MPTCP-next ( https://github.com/multipath-tcp/mptcp_net-next ) not yet upstreamed.

For OpenMPTCProuter, clients connexions must be able to join VPS public IP and VPS and router use patched MPTCP kernel. 
MPTCP v0 is used from https://github.com/multipath-tcp/mptcp/tree/mptcp_trunk with some patches and updated to kernel 5.4.10*.
Between router and VPS a proxy is used for TCP (Shadowsocks or V2ray that transport UDP over TCP so work for UDP too), and a VPN for traffic that is not TCP (ICMP, UDP,...).
